rpcs3/Emu/Cell/Modules/cellRtc.cpp

```diff
@@ -3,45 +3,160 @@
 
 #include "cellRtc.h"
 
+#include "Emu/Cell/lv2/sys_time.h"
+#include "Emu/Cell/lv2/sys_memory.h"
+#include "Emu/Cell/lv2/sys_ss.h"
+
 LOG_CHANNEL(cellRtc);
 
+const u64 EPOCH_AS_FILETIME = 116444736000000000;
+const u64 UTC_BASE_TICKS    = 62135596800000000;
+
+// Also stores leap year
+const u8 days_in_month[24] = {0x1F, 0x1C, 0x1F, 0x1E, 0x1F, 0x1E, 0x1F, 0x1F, 0x1E, 0x1F, 0x1E, 0x1F, 0x1F, 0x1D, 0x1F, 0x1E, 0x1F, 0x1E, 0x1F, 0x1F, 0x1E, 0x1F, 0x1E, 0x1F};
+
 s64 convertToUNIXTime(u16 seconds, u16 minutes, u16 hours, u16 days, s32 years)
 {
-	return s64{seconds} + s64{minutes} * 60 + s64{hours} * 3600 + s64{days} * 86400 +
-		s64{years - 70} * 31536000 + s64{(years - 69) / 4} * 86400 -
-		s64{(years - 1) / 100} * 86400 + s64{(years + 299) / 400} * 86400;
+	return s64{seconds} + s64{minutes} * 60 + s64{hours} * 3600 + s64{days} * 86400 + s64{years - 70} * 31536000 + s64{(years - 69) / 4} * 86400 - s64{(years - 1) / 100} * 86400 +
+	       s64{(years + 299) / 400} * 86400;
 }
 
 u64 convertToWin32FILETIME(u16 seconds, u16 minutes, u16 hours, u16 days, s32 years)
 {
-	s64 unixtime = convertToUNIXTime(seconds, minutes, hours, days, years);
-	u64 win32time = static_cast<u64>(unixtime) * 10000000 + 116444736000000000;
+	s64 unixtime      = convertToUNIXTime(seconds, minutes, hours, days, years);
+	u64 win32time     = static_cast<u64>(unixtime) * 10000000 + EPOCH_AS_FILETIME;
 	u64 win32filetime = win32time | win32time >> 32;
 	return win32filetime;
 }
 
-s32 cellRtcGetCurrentTick(vm::ptr<CellRtcTick> pTick)
+// TODO make all this use tick resolution and figure out the magic numbers
+
+// Internal helper functions
+
+s32 setSecureRtcTime(u64 time)
+{
+	return sys_ss_secure_rtc(0x3003 /* SET_TIME */, time, 0, 0); //	TODO
+}
+
+s32 getSecureRtcTime(u64 unk1, u64 unk2, u64 unk3)
+{
+	return sys_ss_secure_rtc(0x3002 /* GET_TIME */, unk1, unk2, unk3); //	TODO
+}
+
+// End of internal helper functions
+
+error_code cellRtcGetCurrentTick(vm::ptr<CellRtcTick> pTick)
 {
 	cellRtc.todo("cellRtcGetCurrentTick(pTick=*0x%x)", pTick);
 
+	error_code ret = sys_memory_get_page_attribute(pTick.addr(), vm::make_var<sys_page_attr_t>({}));
+	if (ret != CELL_OK)
+	{
+		return CELL_RTC_ERROR_INVALID_POINTER;
+	}
+
+	vm::var<s64> sec  = 0;
+	vm::var<s64> nsec = 0;
+
+	ret = sys_time_get_current_time(sec, nsec);
+	if (ret != CELL_OK)
+	{
+		return ret;
+	}
+
+	pTick->tick = *nsec / 1000 + *sec * cellRtcGetTickResolution() + UTC_BASE_TICKS;
+
 	return CELL_OK;
 }
 
 s32 cellRtcGetCurrentClock(vm::ptr<CellRtcDateTime> pClock, s32 iTimeZone)
 {
 	cellRtc.todo("cellRtcGetCurrentClock(pClock=*0x%x, time_zone=%d)", pClock, iTimeZone);
 
+	vm::ptr<sys_page_attr_t> page_attr = vm::make_var<sys_page_attr_t>({});
```

Comment:
```suggestion
	vm::var<sys_page_attr_t> page_attr({});
```
This is not just a cosmetic change, the previous code was bugged.
And in other places.

Reply:
Done